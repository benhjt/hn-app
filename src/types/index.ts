export type StoryType = {
  id: number;
  title: string;
  points: number;
  content: string;
  user: string;
  time: number;
  time_ago: string;
  type: string;
  url: string;
  domain: string;
  comments_count: number;
  comments?: Array<CommentType>;
};

export type CommentType = {
  id: number;
  level: number;
  user: string;
  time: number;
  time_ago: string;
  content: string;
  comments?: Array<CommentType>;
};

export type UserType = {
  id: string;
  created: number;
  karma: number;
};
