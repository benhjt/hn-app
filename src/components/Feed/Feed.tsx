import { Link } from 'react-router-dom';
import { FeedItem } from '../FeedItem';
import { StoryType } from '../../types';

const Feed = ({
  type,
  feed,
  page,
}: {
  type: string;
  feed: StoryType[];
  page: string | undefined;
}) => {
  const nextPage = (Number(page) || 0) + 1;
  return (
    <>
      <div className="flex flex-col gap-2">
        {feed.map((item: StoryType, index: number) => (
          <FeedItem item={item} index={index} key={item.id} />
        ))}
      </div>
      <Link to={`/${type}/${nextPage}`} className="hover:underline">
        More
      </Link>
    </>
  );
};

export default Feed;
