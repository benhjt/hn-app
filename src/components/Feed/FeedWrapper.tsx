import { useParams } from 'react-router-dom';

import Spinner from '../Spinner';
import { StoryType } from '../../types';
import useFetch from 'react-fetch-hook';
import Feed from './Feed';

const FeedWrapper = ({ type }: { type: string }) => {
  const { page } = useParams<'page'>();
  const { isLoading, data, error } = useFetch<StoryType[]>(
    `https://node-hnapi.herokuapp.com/${type}?page=${page}`,
  );

  if (isLoading || !data) return <Spinner />;

  if (error) return <div>Oh no ðŸ˜¢</div>;

  return <Feed type={type} feed={data} page={page} />;
};

export default FeedWrapper;
