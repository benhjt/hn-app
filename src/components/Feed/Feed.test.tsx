import { render } from '@testing-library/react';

import { BrowserRouter as Router } from 'react-router-dom';

import Feed from './Feed';
import { StoryType } from '../../types';

const feed: StoryType[] = [
  {
    id: 37898999,
    title: 'Better HTTP server routing in Go 1.22',
    points: 173,
    user: 'ingve',
    time: 1697459894,
    time_ago: '4 hours ago',
    type: 'link',
    url: 'https://eli.thegreenplace.net/2023/better-http-server-routing-in-go-122/',
    domain: 'eli.thegreenplace.net',
    comments: [],
    comments_count: 97,
    content: '',
  },
];

describe('<Feed />', () => {
  it('renders correctly', () => {
    const { container } = render(
      <Router>
        <Feed feed={feed} type="news" page="1" />
      </Router>,
    );
    expect(container).toMatchSnapshot();
  });
});
