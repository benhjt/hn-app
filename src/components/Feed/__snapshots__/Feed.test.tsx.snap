// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Feed /> renders correctly 1`] = `
<div>
  <div
    class="flex flex-col gap-2"
  >
    <div
      class="flex"
    >
      <div
        class="flex flex-col"
      >
        <div
          class="flex flex-row align space-x-2"
        >
          <a
            class="peer visited:text-gray-500"
            href="https://eli.thegreenplace.net/2023/better-http-server-routing-in-go-122/"
            rel="noopener noreferrer"
            target="_blank"
          >
            Better HTTP server routing in Go 1.22
          </a>
          <a
            class="peer-visited:text-gray-500"
            href="eli.thegreenplace.net"
            rel="noopener noreferrer"
            target="_blank"
          >
            (
            eli.thegreenplace.net
            )
          </a>
        </div>
        <div
          class="flex flex-row text-gray-500 text-xs"
        >
          173
           points by 
          <a
            class="hover:underline"
            href="/user/ingve"
          >
            ingve
          </a>
           
          4 hours ago
          <div
            class="px-2"
          >
            |
          </div>
          <a
            class="hover:underline"
            href="/story/37898999"
          >
            97
             comments
          </a>
        </div>
      </div>
    </div>
  </div>
  <a
    class="hover:underline"
    href="/news/2"
  >
    More
  </a>
</div>
`;
